games-emulation/retroarch/retroarch-1.7.1.recipe

```diff
@@ -0,0 +1,89 @@
+SUMMARY="The reference frontend of libretro"
+DESCRIPTION="RetroArch is a frontend for emulators, game engines and media \
+players. It enables you to run classic games on a wide range of computers and \
+consoles through its slick graphical interface. It has advanced features like \
+shaders, netplay, rewinding, next-frame response times, and more!"
+HOMEPAGE="https://libretro.com/"
+COPYRIGHT="2010-2018 The RetroArch Team"
+LICENSE="GNU GPL v3"
+REVISION="1"
+SOURCE_URI="https://github.com/libretro/RetroArch/archive/v$portVersion.tar.gz"
+CHECKSUM_SHA256="e9189c7da6b2ca0142528eca14f26104ee322f052f8609203d90b60938d8d3f3"
+SOURCE_FILENAME="retroarch-$portVersion.tar.gz"
+SOURCE_DIR="RetroArch-$portVersion"
+
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
+SECONDARY_ARCHITECTURES="x86"
+
+GLOBAL_WRITABLE_FILES="
+	$relativeSettingsDir/retroarch.cfg auto-merge
+	"
+
+USER_SETTINGS_FILES="
+	$relativeSettingsDir/retroarch/retroarch.cfg template \
+		$sysconfDir/retroarch.cfg
+	"
+
+PROVIDES="
+	retroarch$secondaryArchSuffix = $portVersion
+	app:retroarch$secondaryArchSuffix = $portVersion
+	cmd:retroarch$secondaryArchSuffix = $portVersion
+	cmd:retroarch_cg2glsl$secondaryArchSuffix = $portVersion
+	"
+REQUIRES="
+	haiku$secondaryArchSuffix
+	lib:libsdl2_2.0$secondaryArchSuffix
+	lib:libGL$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	lib:libavcodec$secondaryArchSuffix
+	lib:libavformat$secondaryArchSuffix
+	lib:libavdevice$secondaryArchSuffix
+	lib:libswresample$secondaryArchSuffix
+	lib:libavresample$secondaryArchSuffix
+	lib:libavutil$secondaryArchSuffix
+	lib:libswscale$secondaryArchSuffix
+	retroarch_assets
+	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libsdl2$secondaryArchSuffix
+	devel:libgl$secondaryArchSuffix
+	devel:libz$secondaryArchSuffix
+	"
+BUILD_PREREQUIRES="
+	cmd:gcc$secondaryArchSuffix
+	cmd:make
+	cmd:pkg_config$secondaryArchSuffix
+	cmd:which
+	"
+
+PATCH()
+{
+	libretroDir="`finddir B_SYSTEM_DIRECTORY`/$relativeAddOnsDir/libretro"
+	retroarchDataDir="`finddir B_SYSTEM_DATA_DIRECTORY`/retroarch"
+	sed -i \
+		-e "s|^# \(libretro_directory =\)|\1 \"$libretroDir/\"|;" \
+		-e "s|^# \(libretro_info_path =\)|\1 \"$libretroDir/\"|;" \
+		-e "s|^# \(content_directory =\)|\1 \"$retroarchDataDir/\"|;" \
```

Comment:
I've just managed to use the _"Online Updater"_ → _"Content Downloader"_ menu, but I had to change the `content_directory` in my `retroarch.cfg`. So I guess we should change `PATCH()` a little. Here's a suggestion:
```diff
 	libretroDir="`finddir B_SYSTEM_DIRECTORY`/$relativeAddOnsDir/libretro"
-	retroarchDataDir="`finddir B_SYSTEM_DATA_DIRECTORY`/retroarch"
+	assetsDir="`finddir B_SYSTEM_DATA_DIRECTORY`/retroarch/assets"
+	contentDir="`finddir B_USER_DATA_DIRECTORY`/retroarch"
 	sed -i \
 		-e "s|^# \(libretro_directory =\)|\1 \"$libretroDir/\"|;" \
 		-e "s|^# \(libretro_info_path =\)|\1 \"$libretroDir/\"|;" \
-		-e "s|^# \(content_directory =\)|\1 \"$retroarchDataDir/\"|;" \
-		-e "s|^# \(assets_directory =\)|\1 \"$retroarchDataDir/assets/\"|;" \
+		-e "s|^# \(assets_directory =\)|\1 \"$assetsDir/\"|;" \
+		-e "s|^# \(content_directory =\)|\1 \"$contentDir/\"|;" \
 		retroarch.cfg
```
And if retroarch does not create the content directory itself (either at startup or when the user chooses to download something) then we will also need to add a post-install action to create it.

Reply:
A small concern of mine regarding the use of `finddir` for global directories is that the resulting package won't use the correct resources when installed per-user. The assets folder then would be `/home/user/config/data/retroarch/assets`, and not `/system/data/retroarch/assets`, which is hard-coded into the configs.

It's possible to install a package per-user, see [Infrastructure].

Also, as described by [DirectoryStructure], `/home/user/config/data` is not writable, so that's something to keep in mind.

[Infrastructure]: https://git.haiku-os.org/haiku/tree/docs/develop/packages/Infrastructure.rst#n60
[DirectoryStructure]: https://git.haiku-os.org/haiku/tree/docs/develop/packages/DirectoryStructure.rst

Reply:
Good catch, thanks!
Shall we use
```diff
-	contentDir="`finddir B_USER_DATA_DIRECTORY`/retroarch"
+	contentDir="`finddir B_SYSTEM_NONPACKAGED_DATA_DIRECTORY`/retroarch"
```
This would result in `contentsDir=/boot/system/non-packaged/data/retroarch/`.

BTW, the assets directory would still be `/boot/system/data/retroarch/assets/`.

Reply:
Wouldn't this defeat the purpose though? Isn't the point of packaging to not use /non-packaged/?

Reply:
(honest question, maybe there's a meaning behind /non-packaged/ that I am not aware of)

Reply:
Here the `contentDir` is just a path that we are storing in the default config.
The build process (of the retroarch recipe) does not create anything in there.
If we used `finddir B_SYSTEM_DATA_DIRECTORY`/retroarch"` for the `content_directory` config parameter then retroarch would be unable to write anything in there (as this directory is read-only.)

Reply:
I see your point. It makes sense to configure the content directory to a writable location for users to download things through the online updater.

I have the feeling though that a per-user package installation is a very specific use case, where it makes way more sense for the user to modify directly the retroarch.cfg file created upon the first run of retroarch than to use the default paths.

Reply:
I think I have a nice alternative:
```
	contentDir="`finddir B_SYSTEM_VAR_DIRECTORY`/retroarch"
```
This has several advantages:
* It is writable :)
* `finddir B_SYSTEM_VAR_DIRECTORY` yelds `/boot/system/var` and this is common to all users.
* An `mkdir -p $sharedStateDir/retroarch` will create the empty directory at installation time.
* If that directory already exists before installation then its contents are kept.

(I'm not sure yet about the last two items, so I'll give it a try.)

Reply:
Cool! Waiting for your feedback before I push a new commit then.

Reply:
I have updated my comment [here](https://github.com/haikuports/haikuports/pull/2260#discussion_r172115179).
With these settings I can  choose _Online Updater_ → _Content Downloader_ and see my downloads land in `/system/var/retroarch/assets/`.
I still have to check whether or not we need a post-install to create some directories.

BTW, I haven't investigated yet the topic about the template config. I'll do that tomorrow ;-)

Regading the `SUPPLEMENTS` I did a test and I can confirm that pkgman does not allow for a fine grained selection. If the user asks to install retroarch, then it does ask for confirmation before install... but there is not way to say we only want to install retroarch. Maybe we should add a `--ignore-supplements` to pkgman ;)

Reply:
@fbrosson I think `--ignore-supplements` should be added and be the default. This really reminded me about `apt-get` family where I always have to pass `--no-install-recommends` to avoid unneeded packages.

Reply:
I pushed a commit with your changes @fbrosson - thanks!

Reply:
What a coincidence, @alaviss! I also don't like to install packages that are only recommended.
So, in case you did not find yet a workaround for the `--no-install-recommends`, here's how I "fixed" this issue. I just created a file named `/etc/apt/apt.conf.d/no-recommended.conf` and placed this line in it:
```
Apt::Install-Recommends "false";
```

Reply:
You are mistaken about what "supplements" is. The goal is not to provide some optional extension to a package. The goal is to do a reverse "requires".

This is used for example for the qthaikuplugins package, which provides the rendering engine for Qt. Without this package it is not possible to use Qt at all. However, it is a plug-in and could be replaced, so Qt does not know that it depends on it.

From the documentation of "supplements":

    this package will automatically be selected for installation if the supplemented entities are already installed.

The use of this is quite rare, usually software is built with known dependencies and REQUIRES can be used and works fine. Only in some rare cases a SUPPLEMENTS is needed. But when it is needed, the relationship is as strong as REQUIRES, and not something you should be able to disable with a command line flag.

We don't have anything similar to apt "recommends" scheme, and we don't plan to add such a thing, because no one likes it :)

Reply:
What you describe with plug-ins has already been done, for example for some times we had mesa depending on either mesa_swrast or mesa_swpipe renderers. Using REQUIRES with an agreed upon dependency name works just fine here.

Back to the case at hand: to me it makes perfect sense that the retroarch plugins require the core. What would you do with the plugins if the core is not installed? There is a dependency, just like if retroarch was a library and all the cores would be executables depending on it. It just happens to use different technical choices, making the shared code an executable and the specific one libraries. But that is irrelevant to the package manager. For any engine to be usable you need retroarch to be installed. So they should depend on it. If that was not the case, we would have spared ourselves the effort of manually writing REQUIRES/SUPPLEMENTS at all and would let the package manager guess from executable and libraries inside the packages directly. We added a separate dependency system precisely because this wasn't flexible enough to cover some unusual cases.

Reply:
> to me it makes perfect sense that the retroarch plugins require the core. What would you do with the plugins if the core is not installed?

I think what you mean by "core" is "engine". In libretro speak, a "core" is what we are calling a plugin.

Anyway, the fact that a plugin is useless without an engine does not mean that the plugin itself requires the engine. It's the **_user_** who needs the engine, not the plugin.

That said, I'm OK with having a `REQUIRES="retroarch"` in each plugin.

I only wish the implementation of `SUPPLEMENTS` was a bit more flexible so that we would be able to replace the `REQUIRES="retroarch"` by a `SUPPLEMENTS="retroarch"` without having all the plugins installed when the user installs retroarch. But I can live with this :)